Add a BuiltinGenericType base class.

jckarter opened the pull request:
Builtin.FixedArray was introduced as the first generic builtin type, with special case handling in all the various recursive visitors. Introduce a base class, and move the handling to that base class, so it is easier to introduce other generic builtins in the future.

jckarter committed:
Add a BuiltinGenericType base class.

Builtin.FixedArray was introduced as the first generic builtin type, with
special case handling in all the various recursive visitors. Introduce
a base class, and move the handling to that base class, so it is easier
to introduce other generic builtins in the future.

jckarter commented on lib/AST/Builtins.cpp:
```diff
@@ -3663,62 +3664,113 @@ BuiltinUnboundGenericType::getGenericSignature() const {
+
+GenericSignature
+BuiltinGenericType::getGenericSignature() const {
+  return getBuiltinGenericSignature(getASTContext(), getKind());
```
Well, `getSubstitutions()` is implemented in terms of `getGenericSignature`, but caching it sounds like a good idea.

slavapestov commented on lib/AST/Builtins.cpp:
```diff
@@ -3663,62 +3664,113 @@ BuiltinUnboundGenericType::getGenericSignature() const {
+                           SubstitutionMap subs,
+                           bool validate) {
+  if (!subs.getGenericSignature()->isEqual(getBuiltinGenericSignature(C, kind))) {
+    return ErrorType::get(BuiltinUnboundGenericType::get(kind, C));
```
(It would also be cleaner if this method always returned the right type of Type, instead of just Type itself.)

jckarter commented:
@slavapestov I added caches and removed the assertion from CSSimplify; how's this look now?

I can look into chasing down the ErrorType clients in a follow-up.

slavapestov reviewed (approved):
Alright, sounds good. Yeah we generally want to form ErrorTypes as close as possible "to the surface" because when they propagate around silently it can be difficult to debug the root cause of the problem.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test Windows
